### PR TITLE
Use pre-commit

### PR DESCRIPTION
**Pull request by blink1073**
## References
First step of #12272 

## Code changes
- Switches from `husky` to `pre-commit` while running the same checks and same local user experience (auto-install git hooks).
- Adds auto-formatting and checks used by `jupyter_server` (except flake8 which will follow).
- Runs the autoformatter as a separate commit that can be ignored 
- Updates contributing guide

## User-facing changes
None

## Backwards-incompatible changes
None

### CODE REVIEW

**Comment by blink1073** [comment]
That was easy.  😅

File: setup.py [hunk 1]
```diff
@@ -1,66 +1,85 @@
...
-# Distributed under the terms of the Modified BSD License.
-from os.path import join as pjoin
 import json
 import os
 import os.path as osp
+import subprocess
 import sys
 
+# Copyright (c) Jupyter Development Team.
+# Distributed under the terms of the Modified BSD License.
```
Comment (fcollonval) [inline setup.py:11]: This got move by mistake I think

**Review by fcollonval — COMMENTED** [review]
Small lines motion with the license notice. LGTM otherwise

File: setup.py (hunk 1 above)
Comment (blink1073) [reply setup.py:11]: Fixed

**Review by fcollonval — APPROVED** [review]
Thanks @blink1073

**Comment by bollwyvl** [comment]
From a downstream perspective, adding `pre-commit` as part of the `build-system` seems super heavy, as noted on https://github.com/conda-forge/jupyterlab-feedstock/pull/333

**Comment by vidartf** [comment]
@blink1073 This PR switched the lockfile from using yarnpkg to using npmjs (https://github.com/jupyterlab/jupyterlab/pull/12742). Was this intentional? If so, we can switch the default. If it was an accident we can revert it, as currently proposed in that PR.

**Comment by blink1073** [comment]
> Was this intentional?

No, it was not intentional

**Pull request merged**